### PR TITLE
cmake: Always add available languages if no language is specified (fixes #6441)

### LINKED ISSUE #6441: regression in 0.53. meson cannot handle CheckCXXSourceRuns.cmake
Hi there,

**Describe the bug**

I'm trying to link with project that uses CMake and it checks supported CPU instructions via [CHECK_CXX_SOURCE_RUNS](https://cmake.org/cmake/help/v3.0/module/CheckCXXSourceRuns.html). Meson 0.52 were handling it without any problems, but 0.53 cannot.

Here is cmake file that causes issue [CPUID.cmake](https://github.com/opencv/dldt/blob/2019/inference-engine/src/extension/cmake/CPUID.cmake)

**To Reproduce**

You should install that project and use this meson.build:

```meson
project('mytestproject', ['c', 'cpp'])

inference_engine_dep = dependency('InferenceEngine', version: '>=2.0.0', required: true, method : 'cmake', modules: ['IE::inference_engine'])
```

**Expected behavior**

After running `meson build/` you must see following output:

```sh
Run-time dependency inferenceengine (modules: IE::inference_engine) found: YES 2.1.0
```

**system parameters**
* This is just a plain native build
* Operating system is Ubuntu 18.04
* Python version is 3.6.9
* meson version is 0.53.0

### PR DESCRIPTION
**Pull request by mensinda**
The default behavior for project() is to add the C and CXX languages. Currently, only the C language is added if both c and cpp languages are used in meson.

### CODE REVIEW

**Comment by scivision** [comment]
I didn't try this myself, but I think this assumes that C++ is always available, which is not always true. 
For example if the machine does not have a C++ compiler or is using a backend where C++ is goofed up I think this PR might make those cases fail.

I wrote this code originally, and I think the original code may have one or two defects:

### for loop instead of if/elif

https://github.com/mesonbuild/meson/pull/6444/files#diff-bc427959078b2a6a3516350d36b823abL1060

where I perhaps should have done a for loop instead of if/elif. I didn't try this.

### C++ implies C also

If we have C++ project, we maybe should assume we have C as well, as was necessary in general for Fortran projects using CMake: https://github.com/mesonbuild/meson/pull/6444/files#diff-bc427959078b2a6a3516350d36b823abL1512

**Comment by mensinda** [comment]
Now that I think more about it, wouldn't it make sense to have a language array instead of a single language variable for CMake? This way, we could enable all supported languages.

**Comment by scivision** [comment]
Do you mean to have `depedency(..., language=['cpp', 'c'])` or internally?

I think in any case, there should be the internal CMake Langauge patching like https://github.com/mesonbuild/meson/pull/6444/files#diff-bc427959078b2a6a3516350d36b823abL1518
since the end-user will not want to have to think about `dependency(..., language='fortran')` also needs to use `C` in CMake

**Comment by mensinda** [comment]
At least internally for now with the additional patching for Fortran. The rough procedure I had in mind was this:

```
if language is None
  add all languages meson knows to language_array
else
  add language to language_array
endif
patch language_array
```

**Comment by scivision** [comment]
Yes I think you're saying that if the user does not specify a language in `dependency()`, then default to all languages of the Meson project, assuming the language is also supported by CMake. That would avoid the case of assuming C++ when it isn't actually available, because Meson would previously have verified C++ is available.

File: mesonbuild/dependencies/base.py [hunk 1]
```diff
@@ -1055,22 +1055,28 @@ def _original_module_name(self, module: str) -> str:
...
+                compilers = environment.coredata.compilers.build
             else:
-                if 'c' in environment.coredata.compilers.host.keys():
-                    language = 'c'
-                elif 'cpp' in environment.coredata.compilers.host.keys():
-                    language = 'cpp'
-                elif 'fortran' in environment.coredata.compilers.host.keys():
-                    language = 'fortran'
+                compilers = environment.coredata.compilers.host
+
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__ blank line contains whitespace ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6444#issue-5063603))

<details>

| Rule |
| - |
| W293 |

</details>
<sub>posted by Sider</sub>

**Comment by mensinda** [comment]
@scivision Updated and the CI is green. Could you have another look?

**Comment by mensinda** [comment]
Ping :)

**Comment by nirbheek** [comment]
We're going to make the 0.53.1 release today, so would be good to get this in if possible.

File: mesonbuild/dependencies/base.py [hunk 2]
```diff
@@ -1055,21 +1055,26 @@ def _original_module_name(self, module: str) -> str:
...
+            candidates = ['c', 'cpp', 'fortran', 'objc', 'objcxx']
+            self.language_list += [x for x in candidates if x in compilers.keys()]
```
Comment (dcbaker) [inline mesonbuild/dependencies/base.py]: this really should just be:
```python
self.language_list += [x for x in candidates if x in compilers]
```
With is log(n) (it's a hash lookup), as opposed to the `keys()` version with is o(n)

**Comment by mensinda** [comment]
I am going to merge this PR once the CI is green and there are no objections.

**Comment by scivision** [comment]
I didn't run this step by step, but it looks like a better approach and I don't think any issues have been noted yet from `master` users. Great!

**Pull request merged**